[MP Rewrite Part 9] Improve test coverage for motion planning primitives

kylestach opened the pull request:
## Description
Adding new automated and unit tests for motion planning primitives.

kylestach committed:
Write Bezier path tests

kylestach committed:
Improve automated testing

kylestach committed:
Fix some primitives for tests

kylestach committed:
Add angle overrides and fix pivot planner. Small tweaks to settle.

JNeiger commented on soccer/planning/tests/AnglePlanningTest.cpp:
```diff
@@ -23,24 +32,73 @@ TEST(AnglePlanning, FaceAngle) {
+    check_angle_planning_near_exact(trajectory, angle_fn, 5e-2);
+}
+
+TEST(AnglePlanning, DISABLED_start_from_incorrect_angle) {
```
Whats the thought behind this being disabled?

kylestach replied:
It is a desired feature of angle planning that is currently not implemented.

JNeiger commented on soccer/planning/tests/BezierPathTest.cpp:
```diff
@@ -6,35 +6,98 @@
+// Smoothed paths should have reasonably low curvature everywhere on the
+// path. Sadly, this does not hold with our current system, which uses
+// Bezier curves and places keypoints in a fairly naive manner.
+// All examples with zero-velocity endpoints are broken because of numerical issues (in this case, the second and third control points go on top of the first and fourth, respectively).
```
Might need to make pretty

kylestach replied:
Yeah I'm doing that in a separate part, it's kind of annoying to run make pretty on each of them individually.

JNeiger commented on soccer/planning/tests/BezierPathTest.cpp:
```diff
@@ -6,35 +6,98 @@
+// path. Sadly, this does not hold with our current system, which uses
+// Bezier curves and places keypoints in a fairly naive manner.
+// All examples with zero-velocity endpoints are broken because of numerical issues (in this case, the second and third control points go on top of the first and fourth, respectively).
+// TODO(Kyle): Switch to a scheme that minimizes sum of squared
```
Is there an issue for this? I haven't checked

kylestach replied:
There is now, #1539.

JNeiger commented on soccer/planning/tests/CreatePathTest.cpp:
```diff
@@ -0,0 +1,3 @@
+#include "planning/low_level/CreatePath.hpp"
+
+#include <gtest/gtest.h>
```
Is this intentional?

kylestach replied:
Yes? What about it?

JNeiger replied:
It was just an empty file with two includes

kylestach replied:
Oh rip, thought I actually put stuff in here. Honestly I think as much as can be tested with this is covered by other tests, I'll delete it.

JNeiger commented on soccer/planning/tests/PlannerTest.cpp:
```diff
@@ -298,8 +300,8 @@ TEST(Planning, settle_random) {
     for (int i = 0; i < 500; i++) {
-        world_state.ball.position = Point{random(-1.5, 1.5), random(2, 4)};
-        world_state.ball.velocity = Point{random(-.3, .3), random(-0.3, -.2)};
+        world_state.ball.position = Point{random(-1.5, 1.5), random(3, 4)};
```
I know this was mentioned in the future, but is it planned to make this deterministic?

From what I can tell, as long as the big compilers are used and the specific prng generator is specified, it should be deterministic across machines. 

Specifically the example here with mt19937 should work if we set a static see
https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution

kylestach replied:
Yeah, setting a deterministic seed is the easy fix.

kylestach committed:
Add TODOs for curvature limiting/Bezier spline issues

kylestach committed:
Make planner tests deterministic

kylestach committed:
Merge branch 'mp-rewrite/automated-tests' into mp-rewrite/final-planner-improvements

kylestach committed:
Clean up and comment logic in settle planner

JNeiger reviewed (approved):
Besides that one empty file with 2 includes, it lgtm

oswinso commented on soccer/planning/tests/PlannerTest.cpp:
```diff
@@ -68,11 +77,11 @@ TEST(Planning, DISABLED_path_target_random) {
 
-    EXPECT_LT(failure_count, 100)
-        << "Failure rate should be less than 10 percent for path target";
+    EXPECT_LT(failure_count, 150)
+        << "Failure rate should be less than 15 percent for path target";
```
Where does this number come from?

kylestach replied:
From my ass.

These tests are garbage and heuristic, it's not really reasonable to put hard expectations on them.

oswinso commented on soccer/planning/tests/PlannerTest.cpp:
```diff
@@ -267,16 +277,17 @@ TEST(Planning, settle_basic) {
     WorldState world_state;
-    world_state.ball.position = Point{1, 1};
-    world_state.ball.velocity = Point{-.1, -.1};
+    // Use some initial velocity, settle doesn't always work for non-moving balls
```
Make an issue?

kylestach replied:
This is fine. Collect is intended to handle those cases.

oswinso reviewed (approved):
lgtm

kylestach committed:
Remove empty CreatePathTest.cpp

kylestach committed:
Re-add RRT tests in CreatePathTest

kylestach committed:
Merge branch 'mp-rewrite/automated-tests' into mp-rewrite/final-planner-improvements

kylestach committed:
Make settle planner pick best buffer

This fixes the edge case where the planner can't find an path that finds
the ball quickly enough because the only way to catch it doesn't have
enough buffer time (i.e. immediately before it is caught).

kylestach committed:
[MP Rewrite Part 11] Miscellaneous formatting and cleanup (#1540)

* Formatting

* More formatting

* Tidy fixes

* Straggling formatting

* Formatting

* Rename low_level to primitives

* Comment on cached_pivot_point

[MP Rewrite Part 12] Port the line kick planner (#1541)

* Add in sketchy line kick

* Formatting

* Tidy

[MP Rewrite Part 13] Stop play fixes and holding trajectory (#1542)

* Fix stopped play and hold in place at end of path move

* Formatting

* Dumb NOLINT

* Style again

kylestach committed:
[MP Rewrite Part 10] Overall improvements to planners (#1537)

* Add angle overrides and fix pivot planner. Small tweaks to settle.

* Clean up and comment logic in settle planner

* Make settle planner pick best buffer

This fixes the edge case where the planner can't find an path that finds
the ball quickly enough because the only way to catch it doesn't have
enough buffer time (i.e. immediately before it is caught).

[MP Rewrite Part 11] Miscellaneous formatting and cleanup (#1540)

* Formatting

* More formatting

* Tidy fixes

* Straggling formatting

* Formatting

* Rename low_level to primitives

* Comment on cached_pivot_point

[MP Rewrite Part 12] Port the line kick planner (#1541)

* Add in sketchy line kick

* Formatting

* Tidy

[MP Rewrite Part 13] Stop play fixes and holding trajectory (#1542)

* Fix stopped play and hold in place at end of path move

* Formatting

* Dumb NOLINT

* Style again
